### PR TITLE
Add support for 2b hashes

### PR DESCRIPTION
**Pull request by agathver**
* Add support for 2b hashes
* Update underlying bcrypt source to improvements in OpenBSD
* Bump node-pre-gyp and nan
* Update README with compatibility note

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -29,11 +29,11 @@
     "install": "node-pre-gyp install --fallback-to-build"
   },
   "dependencies": {
-    "nan": "2.6.2",
-    "node-pre-gyp": "0.6.36"
+    "nan": "^2.7.0",
```
Comment (defunctzombie) [inline package.json]: I _strongly_ recommend you avoid `^` for dependencies. Patch versions have and will break installs.

File: src/bcrypt.cc [hunk 2]
```diff
@@ -229,7 +211,19 @@ bcrypt(const char *key, const char *salt, char *encrypted)
 	/* We dont want the base64 salt but the raw data */
 	decode_base64(csalt, BCRYPT_MAXSALT, (u_int8_t *) salt);
 	salt_len = BCRYPT_MAXSALT;
-	key_len = strlen(key) + (minor >= 'a' ? 1 : 0);
+	if (minor <= 'a')
```
Comment (defunctzombie) [inline src/bcrypt.cc]: use `{`

File: src/bcrypt.cc [hunk 3]
```diff
@@ -229,7 +211,19 @@ bcrypt(const char *key, const char *salt, char *encrypted)
...
+        key_len = (u_int8_t)(strlen(key) + (minor >= 'a' ? 1 : 0));
+    else
```
Comment (defunctzombie) [inline src/bcrypt.cc]: indentation does not match

File: test/async.js [hunk 4]
```diff
@@ -84,7 +84,7 @@ module.exports = {
         assert.expect(2);
         bcrypt.genSalt(10, function(err, salt) {
             var split_salt = salt.split('$');
-            assert.strictEqual(split_salt[1], '2a');
+            assert.strictEqual(split_salt[1], '2b');
```
Comment (defunctzombie) [inline test/async.js]: why are these tests changing?

**Comment by defunctzombie** [comment]
> Add support for 2b hashes

Why is this a desired feature to support?

File: src/bcrypt.cc (hunk 2 above)
Comment (agathver) [reply src/bcrypt.cc]: It's OpenBSD source code, mostly modified. They prefer tabs and miss braces. Changing them will make further diffing with their source code impossible, as we have certain NodeJs and windows specific changes

File: test/async.js (hunk 4 above)
Comment (agathver) [reply test/async.js]: The default becomes 2b in place of 2a

**Comment by agathver** [comment]
@defunctzombie 
2a version has a bug, that 2b fixes. It's been default in all other bcrypt implementationd for quite a while.

**Comment by superlukas** [comment]
There is a newer version of bcrypt.c in OpenBSD master. Any specific reason you want the one from 2014? blowfish.c is outdated too by the way.

See:
https://github.com/openbsd/src/blob/master/lib/libc/crypt/bcrypt.c
https://github.com/openbsd/src/blob/master/lib/libc/crypt/blowfish.c

**Comment by agathver** [comment]
@superlukas  The algorithm and code that calculates bcrypt hash is still the same, however, they changed the function names. I really didn't want to change the entire node.bcrypt.js codebase.

**Comment by superlukas** [comment]
@agathver Sorry, I didn't see that encode_salt is gone. I only saw the classic interface at the bottom and assumed it must be backwards compatible then.

**Comment by ncb000gt** [comment]
The fix here is capping the key length yea?

**Comment by agathver** [comment]
@ncb000gt Yes. Since it breaks old (incorrect) hashes, so a prefix was required, so as to keep the old code intact.

**Comment by agathver** [comment]
We can mark bcrypt v1.x as deprecated with this warning. A run-time warning may be too annoying.

(however, we don't enable any workaround for $2a$ hashes)

**Comment by JessieAMorris** [comment]
What's the word on getting this merged?

**Pull request merged**